components/prism-sas.js

```diff
@@ -1,34 +1,173 @@
-Prism.languages.sas = {
-	'datalines': {
-		pattern: /^\s*(?:(?:data)?lines|cards);[\s\S]+?(?:\r?\n|\r);/im,
-		alias: 'string',
-		inside: {
-			'keyword': {
-				pattern: /^(\s*)(?:(?:data)?lines|cards)/i,
-				lookbehind: true
-			},
-			'punctuation': /;/
-		}
-	},
-	'comment': [
+(function (Prism) {
+
+	var stringPattern = /(["'])(?:\1\1|(?!\1)[\s\S])*\1(?!\1)/.source;
+
+	var number = /\b(?:\d[\da-f]*x|\d+(?:\.\d+)?(?:e[+-]?\d+)?)\b/i;
+	var numericConstant = {
+		pattern: RegExp(stringPattern + '[bx]'),
+		alias: 'number'
+	};
+
+	var step = {
+		pattern: /(?:proc\s+\w+|quit|run|data(?!\=))\b/i,
+		alias: 'keyword'
+	};
+
+	var comment = [
 		{
-			pattern: /(^\s*|;\s*)\*.*;/m,
+			pattern: /(^\s*|;\s*)\*[^;]*;/m,
 			lookbehind: true
 		},
 		/\/\*[\s\S]+?\*\//
-	],
-	'datetime': {
-		// '1jan2013'd, '9:25:19pm't, '18jan2003:9:27:05am'dt
-		pattern: /'[^']+'(?:dt?|t)\b/i,
-		alias: 'number'
-	},
-	'string': {
-		pattern: /(["'])(?:\1\1|(?!\1)[\s\S])*\1/,
+	];
+
+	var string = {
+		pattern: RegExp(stringPattern),
 		greedy: true
-	},
-	'keyword': /\b(?:data|else|format|if|input|proc\s\w+|quit|run|then|libname|set|output|options)\b/i,
-	// Decimal (1.2e23), hexadecimal (0c1x)
-	'number': /\b(?:[\da-f]+x|\d+(?:\.\d+)?(?:e[+-]?\d+)?)/i,
-	'operator': /\*\*?|\|\|?|!!?|¦¦?|<[>=]?|>[<=]?|[-+\/=&]|[~¬^]=?|\b(?:eq|ne|gt|lt|ge|le|in|not)\b/i,
-	'punctuation': /[$%@.(){}\[\];,\\]/
-};
+	};
+
+	var punctuation = /[$%@.(){}\[\];,\\]/;
+
+
+	Prism.languages.sas = {
+		'datalines': {
+			pattern: /^(\s*)(?:(?:data)?lines|cards);[\s\S]+?^;/im,
+			lookbehind: true,
+			alias: 'string',
+			inside: {
+				'keyword': {
+					pattern: /^(?:(?:data)?lines|cards)/i,
+					lookbehind: true
+				},
+				'punctuation': /;/
+			}
+		},
+		'proc-sql': {
+			pattern: /^proc\s+sql(?:\s+[\w|=]+)?;(?:[\s\S]+?^(?:proc\s+\w+|quit|run|data);|(?:\r?\n|\r)[\s\S]+;)/im,
```

Comment:
Could you please this pattern to me? I don't get this part:

```
(?:
    [\s\S]+?^(?:proc\s+\w+|quit|run|data);
|
    (?:\r?\n|\r)[\s\S]+;
)
```

Does this mean that a `proc sql` statement can be terminated by any other `proc *` statement?
From the doc, I thought that a `proc sql` statement has to be ended with a `quit;`.

Reply:
Yep. That's what it means. From the [doc](https://go.documentation.sas.com/?cdcId=pgmsascdc&cdcVersion=9.4_3.4&docsetId=sqlproc&docsetTarget=n0w2pkrm208upln11i9r4ogwyvow.htm&locale=en)
> When a PROC SQL statement is executed, PROC SQL continues to run until a QUIT statement, a DATA step, or another SAS procedure is executed.

Reply:
And what's the second part then?

Reply:
If it’s a code fragment, and it’s not ended, the regex should still recognize the proc.

Reply:
Could you give an example of a code fragment?

---

Code fragments aside, I usually put embedded languages (like SQL) in their token to not mix languages. Since we already a `step` token in SAS, couldn't we match SQL Code like this?

```js
'sql': {
	// TODO: Support for code fragments
	pattern: /(^proc\s+sql(?:\s+[\w|=]+)?;)[\s\S]+?(?=^(?:proc\s+\w+|quit|run|data);)/im,
	lookbehind: true,
	alias: 'language-sql',
	inside: Prism.languages.sql
}
```

(This requires minor adjustments to `step` to also include `proc\s+sql(?:\s+[\w|=]+)?`.)

Reply:
`proc sql outobs=10;
   title 'Proclib.Payroll';
   title2 'First 10 Rows Only';
   select * from proclib.payroll;
   title;`

[Doc](https://go.documentation.sas.com/?cdcId=pgmsascdc&cdcVersion=9.4_3.4&docsetId=sqlproc&docsetTarget=n1lhnlzhrmrqggn1rz570u89oq2m.htm&locale=en). And actually, I mispoke. SAS does not require that you end your program with "run" or "quit". The quoted code would execute just fine.

Reply:
And sure, we can match SQL that way if it's preferred.

Reply:
> The quoted code would execute just fine.

Which quoted code?

Reply:
Question: What is `title` and `title2`. It's not part of regular SQL but right now we use SQL to highlight it.

We should only highlight SQL statements with SQL, so is it possible to exclude any SAS constructs? Maybe this is an impossible task when looking at code like this.

```sas
proc sql;
   title  'Stores Table';
   title2 'Coordinates of Stores';
   select * from stores;
   title  'Houses Table';
   title2 'Coordinates of Houses';
   select * from houses;
title;
```

([Source](https://go.documentation.sas.com/?cdcId=pgmsascdc&cdcVersion=9.4_3.4&docsetId=sqlproc&docsetTarget=n02w9rre6jk3n5n10zot9n4uqrzj.htm&locale=en))

Reply:
> Which quoted code?

The code you just quoted would run. See also:

```
proc sql outobs=10;
   title 'Proclib.Payroll';
   title2 'First 10 Rows Only';
   select * from proclib.payroll;
   title;
```
You do not need to end a proc statement with run or quit. SAS will operate fine just encountering EOF.

Reply:
Ok. I thought you meant SAS code. Sorry about that.
I added the EOF condition to my [proposed pattern](https://github.com/PrismJS/prism/pull/1981#discussion_r307689351).

Reply:
title and title2 are [global statements](https://go.documentation.sas.com/?cdcId=pgmsascdc&cdcVersion=9.4_3.4&docsetId=lestmtsglobal&docsetTarget=n12y6ohua6d9nxn1q1v8kr63t0e5.htm&locale=en).  

Would it be acceptable to declare Prism.languages.sql as a variable within the IIFE and add a global-statements object to it? Then, in the sql object within SAS, use sasSQL rather than Prism.languages.sql?

```
	var globalStatements = {
		pattern: /((?:^|[\s])=?)(?:catname|checkpoint execute_always|dm|endsas|filename|footnote|%include|libname|%list|lock|missing|options|page|resetline|%run|sasfile|skip|sysecho|title\d?)\b/i,
		lookbehind: true,
		alias: 'keyword'
	};

	var sasSQL = Prism.languages.sql;


	sasSQL['global-statements'] = globalStatements;
```

Reply:
Here you modify the SQL language itself.

How about this approach: If my understanding is correct then only a [handful of SQL statements](https://go.documentation.sas.com/?cdcId=pgmsascdc&cdcVersion=9.4_3.4&docsetId=sqlproc&docsetTarget=n1oihmdy7om5rmn1aorxui3kxizl.htm&locale=en) can be used. We can use this because we now that every line which starts which looks like this `/^[ \t]*(?:select|alter table|...)(?:<string>|[^;"'])+;/im`.

This should be able to find us all SQL statements in a `proc sql` statement. Everything has to be regular SAS or more specific global statements.

So like this, I suppose:

```js
'proc-sql': {
	pattern: /^proc\s+sql(?:\s+[\w|=]+)?;[\s\S]+?(?=^(?:proc\s+\w+|quit|run|data);|(?![\s\S]))/im,
	inside: {
		'sql': {
			pattern: RegExp(/^[ \t]*(?:select|alter table|...)(?:<str>|[^;"'])+;/.source.replace(/<str>/g, stringPattern), 'im'),
			alias: 'language-sql',
			inside: Prism.languages.sql
		},
		...regular_SAS
	}
}
```

Reply:
The documentation might make it seem as though those statements are the only that work, but you can write[ fairly complex SQL queries](https://go.documentation.sas.com/?cdcId=pgmsascdc&cdcVersion=9.4_3.4&docsetId=sqlproc&docsetTarget=n0340vzqamcsopn1oo4lvwkbkxuq.htm&locale=en#n0340vzqamcsopn1oo4lvwkbkxuq) within proc sql:

```
libname proclib 'SAS-library';
proc sql;
   create view proclib.jobs(pw=red) as
   select Jobcode,
          count(jobcode) as number label='Number',
          avg(int((today()-birth)/365.25)) as avgage
             format=2. label='Average Age',
          avg(salary) as avgsal
             format=dollar8. label='Average Salary'
   from payroll
   group by jobcode
   having avgage ge 30;
   title1 'Current Summary Information for Each Job Category';
   title2 'Average Age Greater Than or Equal to 30';
   select * from proclib.jobs(pw=red);
title;
```

I've been working on this and the other two proc that allow programmers to execute code from another programming language ([proc groovy](https://go.documentation.sas.com/?cdcId=pgmsascdc&cdcVersion=9.4_3.4&docsetId=proc&docsetTarget=n12njw2j3tuptnn1bmmr0cl857d2.htm&locale=en) and [proc lua](https://go.documentation.sas.com/?cdcId=pgmsascdc&cdcVersion=9.4_3.4&docsetId=proc&docsetTarget=n1w8nl91tml15dn1mw9p5l8oj6hy.htm&locale=en)) using a similar approach to what I described above. You're right, however, I was neglecting the fact I was pointing to and adding to the SQL language support itself. With a bit of experimentation, I hit on:

`var sasSQL = Object.assign({}, Prism.languages.sql);`

That will copy the original language, rather than simply pointing to it. Would that be acceptable?

Reply:
[`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) is ES6, so we can't use it. But you can use `Prism.util.clone` which will make a deep copy of a given object.

However, I still believe that my approach is more desirable. Also, your example of a fairly complex SQL query is matched correctly by my pattern.

---

Groovy and Lua seem to be quite easy to add but let's please do that after SQL.

Reply:
This is what I've got right now based on your suggestion above, but I'm definitely not getting full highlighting
```
		'proc-sql': {
			pattern: /^proc\s+(fed)?sql(?:\s+[\w|=]+)?;[\s\S]+?(?=^(?:proc\s+\w+|quit|run|data);|(?![\s\S]))/im,
			inside: {
				'sql': {
					pattern: RegExp(/^[ \t]*(?:select(?:\s+(distinct|unique))?|alter\s+table|(create|describe|drop)\s+(index|table|view)|insert\s+into|update)(?:<str>|[^;"'])+;/.source.replace(/<str>/g, stringPattern), 'im'),
					alias: 'language-sql',
					inside: Prism.languages.sql
				},
				'global-statements': globalStatements,
				'sqlStatements': {
					pattern: /((?:^|[\s]))(?:disconnect\s+from|exec(ute)?|reset|validate)\b/i,
					lookbehind: true,
					alias: 'keyword'
				}
			}
		},
```
![image](https://user-images.githubusercontent.com/10553110/62301135-01a7df00-b446-11e9-9c2c-c63744234445.png)

I suspect it's due to the fact that "create view" does not end with a semicolon for several lines?

Reply:
Don't use capturing groups if not necessary.

The inserted string pattern references the first capturing group which in your case is `(distinct|unique)`. Making all groups non-capturing fixes the issue.
Also: You can remove `(?:\s+(distinct|unique))?` because it can also be matched by the following `(?:<str>|[^;"'])+`.

All in all, this gives us:

```js
'sql': {
	pattern: RegExp(/^[ \t]*(?:select|alter\s+table|(?:create|describe|drop)\s+(?:index|table|view)|create\s+unique\s+index|insert\s+into|update)(?:<str>|[^;"'])+;/.source.replace(/<str>/g, stringPattern), 'im'),
	alias: 'language-sql',
	inside: Prism.languages.sql
},
```

(I also added support for `CREATE UNIQUE INDEX ...`)

(Also, if you show an example, please provide a link to the example or the code itself because I have to copy the whole thing by hand if you don't...)

Reply:
Sorry about the screenshot of the source. In any case, I've committed changes, and I included that code in the sql test.